Libraries: Add support for TTL servos, based on Robotis Dynamixel 1.0 Protocol

teoPiD opened the pull request:
This PR adds support for TTL servos which are based on Robotis Dynamixel 1.0 communication protocol. 

This work is based on the work developed by tridge in https://github.com/tridge/ardupilot/tree/pr-robotis-servo . In the future, both this Library as RobotisServo should be joined in a single library for TTLServos, with a parameter to choose which communication protocol to use or with a bitmask for each channel with a communication protocol. I recommend this because it would simplify the code.

The code was tested using copter 4.0.7 but couldn't be tested on master (master seems to be bricked and crashed my Pixhawk 1). The code was tested on a Feetech RC TTL servo (I don't have a dynamixel servo).

Although it works in the same way as tridge code, but for protocol 1.0, extra parameters were added, namely the registers address can be set. This enables it to be used with servos with different register addresses. Also, a bitmask was added for the servo channel, which means that the servos can be used in parallel or daisy-chained. It also supports 32 servo channels and was removed the association between SRV ID 0 to Servo ID 1.

peterbarker commented on libraries/AP_SerialManager/AP_SerialManager.cpp:
```diff
@@ -178,7 +178,7 @@ const AP_Param::GroupInfo AP_SerialManager::var_info[] = {
     // @DisplayName: Serial4 protocol selection
     // @Description: Control what protocol Serial4 port should be used for. Note that the Frsky options require external converter hardware. See the wiki for details.
-    // @Values: -1:None, 1:MAVLink1, 2:MAVLink2, 3:Frsky D, 4:Frsky SPort, 5:GPS, 7:Alexmos Gimbal Serial, 8:SToRM32 Gimbal Serial, 9:Rangefinder, 10:FrSky SPort Passthrough (OpenTX), 11:Lidar360, 13:Beacon, 14:Volz servo out, 15:SBus servo out, 16:ESC Telemetry, 17:Devo Telemetry, 18:OpticalFlow, 19:RobotisServo, 20:NMEA Output, 21:WindVane, 22:SLCAN, 23:RCIN, 24:MegaSquirt EFI, 25:LTM, 26:RunCa ...
+    // @Values: -1:None, 1:MAVLink1, 2:MAVLink2, 3:Frsky D, 4:Frsky SPort, 5:GPS, 7:Alexmos Gimbal Serial, 8:SToRM32 Gimbal Serial, 9:Rangefinder, 10:FrSky SPort Passthrough (OpenTX), 11:Lidar360, 13:Beacon, 14:Volz servo out, 15:SBus servo out, 16:ESC Telemetry, 17:Devo Telemetry, 18:OpticalFlow, 19:RobotisServo, 20:NMEA Output, 21:WindVane, 22:SLCAN, 23:RCIN, 24:MegaSquirt EFI, 25:LTM, 26:RunCa ...
```
Argh, sorry, 38 is "reserved" for an incoming ESC :-(

We really do need to mark that somehow.

teoPiD replied:
40 then?
I thought it was weird that the values went from 37 to 39, should have thought about that possibility.

teoPiD replied:
Changed to 40

teoPiD replied:
I have just verified that 40 is now used for AIS. Should I use 41? Or any other?

peterbarker commented on libraries/AP_TTLServo/AP_TTLServo.cpp:
```diff
@@ -0,0 +1,388 @@
+};
+
+// Constructor
+AP_TTLServo::AP_TTLServo(void){
+
```
```suggestion
AP_TTLServo::AP_TTLServo(void)
{
```

teoPiD replied:
Changed all the braces on AP_TTLServo.cpp. Now the functions are according to : "Functions definitions: place each brace on its own line. For methods inside a header file, braces can be inline." - followed tridge code here that's why it wasn't according to this rule

peterbarker replied:
For `if` and `while` blocks etc they go on their own line.  Not for functions.  You will find libraries that don't do that but the majority do.

peterbarker commented on libraries/AP_TTLServo/AP_TTLServo.cpp:
```diff
@@ -0,0 +1,388 @@
+}
+
+// Calculate communication protocol CRC (same as Robotis Dynamixel 1.0 protocol CRC)
+uint8_t AP_TTLServo::calculate_crc(uint8_t *txpacket, uint8_t len){
```
```suggestion
uint8_t AP_TTLServo::calculate_crc(uint8_t *txpacket, uint8_t len)
{
```

teoPiD replied:
I have seen your suggestions about the braces, but isn't that going against the wiki? Namely: "Place the open brace on the line preceding the code block; place the close brace on its own line."

peterbarker commented on libraries/AP_TTLServo/AP_TTLServo.cpp:
```diff
@@ -0,0 +1,388 @@
+
+// Use a broadcast ping to find attached servos
+void AP_TTLServo::detect_servos(void){
+    uint8_t txpacket[6];
+
+    txpacket[PKT_ID] = BROADCAST_ID;
+    txpacket[PKT_LENGTH] = 2;
+    txpacket[PKT_INSTRUCTION] = INST_PING;
```
Use a structure.

teoPiD replied:
Using a structure here would mean using a structure in the send_packet function. Should I do that?

peterbarker replied:
Have a look in `AP_Hott_Telem.cpp` for an example.

teoPiD replied:
Done

peterbarker commented on libraries/AP_TTLServo/AP_TTLServo.h:
```diff
@@ -0,0 +1,86 @@
+
+    AP_TTLServo();
+
+    // Do not allow copies
+    AP_TTLServo(const AP_TTLServo &other) = delete;
+    AP_TTLServo &operator = (const AP_TTLServo&) = delete;
```
```suggestion
    // Do not allow copies
    CLASS_NO_COPY(AP_TTLServo);
```

teoPiD replied:
Corrected

peterbarker reviewed (changes requested):
Just a quick style-guide review.

Marking for DevCall to pass this by tridge.

teoPiD commented:
Changed files according to your comments (haven't changed for the structure) and removed the commits that were removing the trailing whitespaces and ordering the position of the values in AP_SerialManager.h

tridge commented:
@teoPiD how is the protocol different from the existing code in libraries/AP_RobotisServo?
If they are similar, can we have an option in the existing library to support the new type?

tridge commented:
@teoPiD if you can update this to instead be parameters on the existing library then I can test with the servos

teoPiD commented:
> @teoPiD how is the protocol different from the existing code in libraries/AP_RobotisServo?
> If they are similar, can we have an option in the existing library to support the new type?

The protocol is different, so that library will not work with this servos (for example, protocol 2.0 header has one more parameter and length uses 2 bytes and not 1 byte, CRC is different, etc). The protocols are described here: https://emanual.robotis.com/docs/en/dxl/protocol1/ and here: https://emanual.robotis.com/docs/en/dxl/protocol2/ . I have submitted this code but was thinking in the future joining the two libraries and using a parameter to choose which protocol to use. I think it would be the best approach.
